File: backend/app/services/task_service.py
import logging
from typing import List, Optional, Tuple
from datetime import datetime
from fastapi import status
from fastapi.exceptions import HTTPException
from sqlalchemy.orm import Query, Session
from sqlalchemy import and_, or_, desc, asc, func

from ..models.task import Task, Comment, TaskPriority
from ..models.user import User
from ..schemas.task import (
    TaskCreate, TaskUpdate, TaskFilter, TaskStatistics, CommentCreate
)


class TaskService:
    """
    Service class for handling task-related business logic.
    
    Attributes:
        db: SQLAlchemy database session
    """

    def __init__(self, db: Session):
        """
        Initialize the TaskService with a database session.
        
        Args:
            db: SQLAlchemy database session
        """
        self.db = db

    def get_task_by_id(self, task_id: int) -> Optional[Task]:
        """
        Retrieve a task by its ID.
        
        Args:
            task_id: The ID of the task to retrieve
            
        Returns:
            Task object if found, None otherwise
        """
        try:
            task: Task | None = self.db.query(Task).filter(Task.id == task_id).first()
            if not task:
                logging.info(f"Task not found (id={task_id})")
                raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Task not found")
            return task
        except Exception as e:
            logging.error(f"Error retrieving task {task_id}: {e}")
            raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))

    def get_tasks(
        self,
        skip: int = 0,
        limit: int = 10,
        filters: Optional[TaskFilter] = None,
        order_by: str = "created_at",
        order_desc: bool = True
    ) -> Tuple[List[Task], int]:
        """
        Retrieve a list of tasks with filtering, pagination, and sorting.
        
        Args:
            skip: Number of records to skip (for pagination)
            limit: Maximum number of records to return
            filters: TaskFilter object with filtering criteria
            order_by: Field to order by
            order_desc: Whether to order in descending order
            
        Returns:
            Tuple of (list of tasks, total count)
        """
        try:
            query: Query[Task] = self.db.query(Task)
        
            # Apply filters
            if filters:
                query = self._apply_filters(query, filters)
        
            # Get total count before pagination
            total: int = query.count()
        
            # Apply ordering
            query = self._apply_ordering(query, order_by, order_desc)
        
            # Apply pagination
            tasks: List[Task] = query.offset(skip).limit(limit).all()
            
            logging.info(f"Retrieved {len(tasks)} tasks {'with filters' if filters else 'without filters'} {'ordered by' if order_by else ''} {'descending' if order_desc else 'ascending'}")

            return tasks, total
        except Exception as e:
            logging.error(f"Error retrieving tasks: {e}")
            raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))

    def create_task(self, task_data: TaskCreate, user_id: int) -> Task:
        """
        Create a new task.
        
        Args:
            task_data: TaskCreate schema with task information
            user_id: ID of the user creating the task
            
        Returns:
            Created Task object
        
        Raises:
            ValueError: If assigned_to user doesn't exist
        """
        try:
            # Validate assigned_to user exists if provided
            if task_data.assigned_to is not None:
                assigned_user: User | None = self.db.query(User).filter(User.id == task_data.assigned_to).first()
                if not assigned_user:
                    logging.error(f"Assigned user not found (id={task_data.assigned_to})")
                    raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="User not found")

            db_task = Task(
                title=task_data.title,
                description=task_data.description,
                due_date=task_data.due_date,
                priority=task_data.priority,
                assigned_to=task_data.assigned_to,
                created_by=user_id
            )

            self.db.add(db_task)
            self.db.commit()
            self.db.refresh(db_task)

            logging.info(f"Task created (id={db_task.id}, by user={user_id})")
            return db_task
        except Exception as e:
            logging.error(f"Error creating task: {e}")
            raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))

    def update_task(self, task_id: int, task_update: TaskUpdate, user_id: int) -> Optional[Task]:
        """
        Update an existing task.
        
        Args:
            task_id: ID of the task to update
            task_update: TaskUpdate schema with updated fields
            user_id: ID of the user making the update
            
        Returns:
            Updated Task object if found, None otherwise
        """
        try:
            db_task = self.get_task_by_id(task_id)
            if not db_task:
                return None

            # Update only provided fields
            update_data = task_update.model_dump(exclude_unset=True)
            for field, value in update_data.items():
                setattr(db_task, field, value)
                
            self.db.commit()
            self.db.refresh(db_task)
            
            logging.info(f"Task updated (id={task_id}, by user={user_id})")
            return db_task
        except Exception as e:
            logging.error(f"Error updating task {task_id}: {e}")
            raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))

    def delete_task(self, task_id: int, user_id: int) -> bool:
        """
        Delete a task by its ID.
        
        Args:
            task_id: ID of the task to delete
            user_id: ID of the user making the deletion
            
        Returns:
            True if task was deleted, False if not found
        """
        try:
            db_task = self.get_task_by_id(task_id)
            if not db_task:
                return False
            
            self.db.delete(db_task)
            self.db.commit()
            
            logging.info(f"Task deleted (id={task_id}) by user {user_id}")
            return True
        except Exception as e:
            logging.error(f"Error deleting task {task_id}: {e}")
            raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))

    def get_task_statistics(self) -> TaskStatistics:
        """
        Get comprehensive task statistics.
        
        Returns:
            TaskStatistics object with various metrics
        """
        try:
            total_tasks: int = self.db.query(Task).count()
            completed_tasks: int = self.db.query(Task).filter(Task.completed == True).count()
            pending_tasks: int = total_tasks - completed_tasks
            
            # Get overdue tasks (past due date and not completed)
            now = datetime.utcnow()
            overdue_tasks: int = self.db.query(Task).filter(
                and_(
                    Task.due_date < now,
                    Task.completed == False
                )
            ).count()
        
            # Get tasks by priority
            high_priority_tasks: int = self.db.query(Task).filter(
                Task.priority == TaskPriority.HIGH
            ).count()
            medium_priority_tasks: int = self.db.query(Task).filter(
                Task.priority == TaskPriority.MEDIUM
            ).count()
            low_priority_tasks: int = self.db.query(Task).filter(
                Task.priority == TaskPriority.LOW
            ).count()
            
            logging.info(f"Task statistics retrieved: {total_tasks} tasks, {completed_tasks} completed, {pending_tasks} pending, {overdue_tasks} overdue, {high_priority_tasks} high priority, {medium_priority_tasks} medium priority, {low_priority_tasks} low priority")
            
            return TaskStatistics(
                total_tasks=total_tasks,
                completed_tasks=completed_tasks,
                pending_tasks=pending_tasks,
                overdue_tasks=overdue_tasks,
                high_priority_tasks=high_priority_tasks,
                medium_priority_tasks=medium_priority_tasks,
                low_priority_tasks=low_priority_tasks
            )
        except Exception as e:
            logging.error(f"Error getting task statistics: {e}")
            raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))

    def create_comment(self, comment_data: CommentCreate, task_id: int, user_id: int) -> Comment:
        """
        Create a new comment for a task.
        
        Args:
            comment_data: CommentCreate schema with comment information
            task_id: ID of the task to comment on
            user_id: ID of the user creating the comment
            
        Returns:
            Created Comment object
        """
        try:
            db_comment = Comment(
                content=comment_data.content,
                task_id=task_id,
                author_id=user_id
            )
            self.db.add(db_comment)
            self.db.commit()
            self.db.refresh(db_comment)
            
            logging.info(f"Comment added to task {task_id} by user {user_id}")
            return db_comment
        except Exception as e:
            logging.error(f"Error adding comment to task {task_id}: {e}")
            raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))

    def _apply_filters(self, query, filters: TaskFilter) -> Query:
        """
        Apply filters to the task query.
        
        Args:
            query: SQLAlchemy query object
            filters: TaskFilter object with filtering criteria
            
        Returns:
            Filtered query object
        """
        if filters.completed is not None:
            query = query.filter(Task.completed == filters.completed)
        
        if filters.priority is not None:
            query = query.filter(Task.priority == filters.priority)
        
        if filters.assigned_to is not None:
            query = query.filter(Task.assigned_to == filters.assigned_to)
        
        if filters.created_by is not None:
            query = query.filter(Task.created_by == filters.created_by)
        
        if filters.due_before is not None:
            query = query.filter(Task.due_date <= filters.due_before)
        
        if filters.due_after is not None:
            query = query.filter(Task.due_date >= filters.due_after)
        
        if filters.search:
            search_term = f"%{filters.search}%"
            query = query.filter(
                or_(
                    Task.title.ilike(search_term),
                    Task.description.ilike(search_term)
                )
            )
        
        return query

    def _apply_ordering(self, query, order_by: str, order_desc: bool) -> Query:
        """
        Apply ordering to the task query.
        
        Args:
            query: SQLAlchemy query object
            order_by: Field to order by
            order_desc: Whether to order in descending order
            
        Returns:
            Ordered query object
        """
        # Map string field names to Task attributes
        order_fields: dict[str, Task] = {
            "created_at": Task.created_at,
            "updated_at": Task.updated_at,
            "due_date": Task.due_date,
            "priority": Task.priority,
            "title": Task.title
        }
        
        field: Task = order_fields.get(order_by, Task.created_at)
        
        if order_desc:
            query = query.order_by(desc(field))
        else:
            query = query.order_by(asc(field))
        
        return query

File: frontend/services/taskService.ts
"use client";

import { useAppDispatch, useAppSelector } from "@/lib/store/hooks";
import {
  addTask,
  deleteTask,
  markTaskAsCompleted,
  addComment as addCommentAction,
  updateTask,
  setStatistics,
} from "@/lib/store/slices/tasks";
import { TaskFilters } from "@/types/Tasks";
import { AxiosInstance } from "axios";
import { RootState } from "@/lib/store";
import type { AppDispatch } from "@/lib/store";
import { CreateTaskSchema, UpdateTaskSchema } from "@/lib/schemas/tasksSchemas";

/**
 * Service class for fetching and manipulating tasks.
 *
 * This class provides methods to retrieve and create tasks using an Axios instance
 * for HTTP requests. It manages the current task state, pagination, and filters
 * through the Redux store. The task state is utilized to fetch the correct page
 * and size of the task list, and it updates the store with new task data or
 * pagination information upon successful API responses.
 *
 * @example
 * const taskService = new TaskService(axiosInstance);
 * const tasks = await taskService.getTasks();
 * console.log(tasks);
 */
class TaskService {
  /**
   * Axios instance for making HTTP requests.
   */
  private api: AxiosInstance;
  /**
   * Current state of the tasks, including pagination and filters.
   *
   * The task state is obtained from the Redux store using `useAppSelector`.
   * The state is used to determine the correct page and size of the task list
   * when fetching tasks from the API.
   */
  private setTaskState: AppDispatch;
  private csrfToken: string;
  constructor(api: AxiosInstance) {
    this.api = api;
    this.setTaskState = useAppDispatch();
    this.csrfToken = useAppSelector((state: RootState) => state.auth.csrfToken);
  }

  /**
   * Retrieve a list of tasks.
   *
   * @param filters - Optional filters to apply to the task list.
   * @returns Promise that resolves with a list of tasks.
   * @throws Error if there was an error fetching the tasks.
   *
   * Supported filters:
   * - `completed`: boolean to filter by completion status.
   * - `priority`: TaskPriority to filter by priority level.
   * - `assigned_to`: user ID to filter by assigned user.
   * - `created_by`: user ID to filter by creator user.
   * - `due_before`: ISO date string to filter by due date before.
   * - `due_after`: ISO date string to filter by due date after.
   * - `search`: string to search in title and description.
   * - `order_by`: string to order by. Supported fields: `created_at`, `updated_at`, `due_date`, `priority`, `title`.
   * - `order_desc`: boolean to order in descending order.
   */
  async getTasks(filters: TaskFilters = {}, page?: number, size?: number) {
    try {
      const params = {
        ...filters,
        page: page || 1,
        size: size || 10,
      };
      const response = await this.api.get("/tasks/", {
        params,
        headers: {
          "Content-Type": "application/json",
          "no-cache": "true",
        },
      });
      void fetch("/internal/log", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
          level: "info",
          message: "Tasks fetched",
          labels: { channel: "tasks" },
          meta: { ts: Date.now() },
        }),
      }).catch(() => {});
      return response.data;
    } catch (error) {
      void fetch("/internal/log", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
          level: "error",
          message: "Tasks fetching failed",
          labels: { channel: "tasks" },
          meta: { ts: Date.now() },
        }),
      }).catch(() => {});
      console.error("Error fetching tasks:", error);
      throw error;
    }
  }

  /**
   * Fetch task statistics from the server.
   *
   * Retrieves comprehensive statistics including:
   * - Total number of tasks
   * - Number of completed tasks
   * - Number of pending tasks
   * - Number of overdue tasks
   * - Tasks by priority level
   *
   * Updates task state with the fetched statistics.
   *
   * @throws Error if there is an issue fetching task statistics.
   */
  async getTasksStatistics() {
    try {
      const response = await this.api.get("/tasks/statistics", {
        headers: {
          "Content-Type": "application/json",
          "no-cache": "true",
        },
      });
      void fetch("/internal/log", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
          level: "info",
          message: "Tasks statistics fetched",
          labels: { channel: "tasks" },
          meta: { ts: Date.now() },
        }),
      }).catch(() => {});
      this.setTaskState(setStatistics(response.data));
    } catch (error) {
      void fetch("/internal/log", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
          level: "error",
          message: "Tasks statistics fetching failed",
          labels: { channel: "tasks" },
          meta: { ts: Date.now() },
        }),
      }).catch(() => {});
      console.error("Error fetching tasks statistics:", error);
      throw error;
    }
  }

  /**
   * Create a new task.
   *
   * @param task - The task data conforming to CreateTaskSchema to be created.
   * @returns Promise that resolves with the created task data.
   * @throws Error if there was an error creating the task.
   */
  async createTask(task: CreateTaskSchema) {
    try {
      const response = await this.api.post("/tasks/", task, {
        headers: {
          "Content-Type": "application/json",
          "no-cache": "true",
          "x-csrf-token": this.csrfToken,
        },
      });
      void fetch("/internal/log", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
          level: "info",
          message: "Task created",
          labels: { channel: "tasks" },
          meta: { ts: Date.now() },
        }),
      }).catch(() => {});
      this.setTaskState(addTask(response.data));
      return response.status;
    } catch (error) {
      void fetch("/internal/log", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
          level: "error",
          message: "Task creation failed",
          labels: { channel: "tasks" },
          meta: { ts: Date.now() },
        }),
      }).catch(() => {});
      console.error("Error creating task:", error);
      throw error;
    }
  }

  /**
   * Update a task.
   *
   * @param taskId - The ID of the task to be updated.
   * @param task - The task data conforming to UpdateTaskSchema to be updated.
   * @returns Promise that resolves with the updated task data.
   * @throws Error if there was an error updating the task.
   */
  async updateTask(taskId: number, task: UpdateTaskSchema) {
    try {
      const response = await this.api.put(`/tasks/${taskId}`, task, {
        headers: {
          "Content-Type": "application/json",
          "no-cache": "true",
          "x-csrf-token": this.csrfToken,
        },
      });
      void fetch("/internal/log", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
          level: "info",
          message: "Task updated",
          labels: { channel: "tasks" },
          meta: { ts: Date.now() },
        }),
      }).catch(() => {});
      this.setTaskState(updateTask(response.data));
      return response.status;
    } catch (error) {
      void fetch("/internal/log", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
          level: "error",
          message: "Task updating failed",
          labels: { channel: "tasks" },
          meta: { ts: Date.now() },
        }),
      }).catch(() => {});
      console.error("Error updating task:", error);
      throw error;
    }
  }

  /**
   * Marks a task as completed.
   *
   * @param taskId - The ID of the task to be marked as completed.
   * @returns Promise that resolves with the status code of the response.
   * @throws Error if there was an error marking the task as completed.
   */
  async markTaskAsCompleted(taskId: number) {
    try {
      const response = await this.api.put(
        `/tasks/${taskId}`,
        {
          completed: true,
        },
        {
          headers: {
            "Content-Type": "application/json",
            "no-cache": "true",
            "x-csrf-token": this.csrfToken,
          },
        }
      );
      void fetch("/internal/log", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
          level: "info",
          message: "Task marked as completed",
          labels: { channel: "tasks" },
          meta: { ts: Date.now() },
        }),
      }).catch(() => {});
      this.setTaskState(markTaskAsCompleted(taskId));
      return response.status;
    } catch (error) {
      void fetch("/internal/log", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
          level: "error",
          message: "Task marking as completed failed",
          labels: { channel: "tasks" },
          meta: { ts: Date.now() },
        }),
      }).catch(() => {});
      console.error("Error marking task as completed:", error);
      throw error;
    }
  }

  /**
   * Deletes a task.
   *
   * @param taskId - The ID of the task to be deleted.
   * @returns Promise that resolves with the status code of the response.
   * @throws Error if there was an error deleting the task.
   */
  async deleteTask(taskId: number) {
    try {
      const response = await this.api.delete(`/tasks/${taskId}`, {
        headers: {
          "Content-Type": "application/json",
          "no-cache": "true",
          "x-csrf-token": this.csrfToken,
        },
      });
      void fetch("/internal/log", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
          level: "info",
          message: "Task deleted",
          labels: { channel: "tasks" },
          meta: { ts: Date.now() },
        }),
      }).catch(() => {});
      this.setTaskState(deleteTask(taskId));
      return response.status;
    } catch (error) {
      void fetch("/internal/log", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
          level: "error",
          message: "Task deletion failed",
          labels: { channel: "tasks" },
          meta: { ts: Date.now() },
        }),
      }).catch(() => {});
      console.error("Error deleting task:", error);
      throw error;
    }
  }

  /**
   * Adds a comment to a specific task.
   *
   * @param taskId - The ID of the task to which the comment is to be added.
   * @param comment - The content of the comment to be added.
   * @returns Promise that resolves with the status code of the response.
   * @throws Error if there was an error adding the comment.
   */
  async addComment(taskId: number, comment: string) {
    try {
      const response = await this.api.post(
        `/tasks/${taskId}/comments`,
        {
          content: comment,
        },
        {
          headers: {
            "Content-Type": "application/json",
            "no-cache": "true",
            "x-csrf-token": this.csrfToken,
          },
        }
      );
      void fetch("/internal/log", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
          level: "info",
          message: "Comment added to task",
          labels: { channel: "tasks" },
          meta: { ts: Date.now() },
        }),
      }).catch(() => {});
      this.setTaskState(addCommentAction({ taskId, comment: response.data }));
      return response.status;
    } catch (error) {
      void fetch("/internal/log", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
          level: "error",
          message: "Comment adding failed",
          labels: { channel: "tasks" },
          meta: { ts: Date.now() },
        }),
      }).catch(() => {});
      console.error("Error adding comment:", error);
      throw error;
    }
  }
}

export default TaskService;
